### PR TITLE
[2.x] Unify the Navigation API

### PR DESCRIPTION
**Pull request by caendesilva**
## Abstract

Taking inspiration from https://github.com/hydephp/develop/pull/1782, to unify the Navigation API

### Roadmap

- [x] https://github.com/hydephp/develop/issues/1790
- [x] https://github.com/hydephp/develop/pull/1565#discussion_r1672031979
- [x] Support extra attributes for navigation items
- [x] Consider merging the sidebar configuration location (wontfix at this time)
     - Tracked in https://github.com/hydephp/develop/issues/1835 
     - (see example in https://github.com/hydephp/develop/pull/1827)

### Changes

- Breaking: The custom navigation item config now uses array inputs 7188b38 
- Breaking: Renames the `hyde.navigation.subdirectories` option to `hyde.navigation.subdirectory_display` c06a475dfde5f8bd0ba40f781b21f3a5dc8c902b
- Feature: We now support setting custom navigation items in the YAML configuration! Fixes `#1790`
- https://github.com/hydephp/develop/pull/1824
- https://github.com/hydephp/develop/pull/1827
- https://github.com/hydephp/develop/pull/1828
- https://github.com/hydephp/develop/pull/1838
- https://github.com/hydephp/develop/pull/1841

### CODE REVIEW

File: config/hyde.php [hunk 1]
```diff
@@ -358,7 +359,7 @@
         // To get started quickly, you can uncomment the defaults here.
         // See the documentation link above for more information.
         'custom' => [
```
Comment (caendesilva) [inline config/hyde.php:361]: Still not super happy with this setting name. Claude suggests `'additional'`, not sure if it's really better.

Comment (caendesilva) [reply config/hyde.php:361]: Entire suggestion by Claude:

Here's an improved version of the navigation configuration to make the structure more intuitive and consistent.

Here are the key changes and improvements:

1.  Changed "custom" to "additional": This better describes the purpose of this section, which is to add menu items that aren't automatically generated from pages.
2.  Changed the structure of "additional" items: Instead of using a custom `NavigationItem::create()` method, we're using a more standard array format. This makes the configuration more consistent with other parts of the config file and easier to read/write in YAML.
3.  Renamed "subdirectories" to "subdirectory_display": This more clearly indicates that this setting controls how subdirectories are displayed in the menu.
4.  Reordered the sections slightly for better logical flow.

```php
'navigation' => [
    // Determines the order of menu items. Lower values appear first.
    'order' => [
        'index' => 0,
        'posts' => 10,
        'docs/index' => 100,
    ],

    // Customize labels for menu items
    'labels' => [
        'index' => 'Home',
        'docs/index' => 'Docs',
    ],

    // Pages to exclude from the navigation menu
    'exclude' => [
        '404',
    ],

    // Additional menu items (internal or external)
    'additional' => [
        // ['destination' => 'https://github.com/hydephp/hyde', 'label' => 'GitHub', 'priority' => 200],
    ],

    // How to display pages in subdirectories: 'dropdown', 'flat', or 'hidden'
    'subdirectory_display' => 'hidden',
],
```

**Comment by caendesilva** [comment]
### RE: The changed config setting

> * Breaking: Renames the `hyde.navigation.subdirectories` option to `hyde.navigation.subdirectory_display` [c06a475](https://github.com/hydephp/develop/commit/c06a475dfde5f8bd0ba40f781b21f3a5dc8c902b)

**Still not super happy with this name.**

> If we want to stick closely to the original meaning while improving clarity, "subdirectory_handling" or "subdirectory_navigation" might be good choices. If we want to be more specific about it being related to the menu structure, "subdirectory_menu_behavior" could work well. https://github.com/hydephp/develop/commit/c06a475dfde5f8bd0ba40f781b21f3a5dc8c902b#commitcomment-144083696

### Open ended AI recommendation:

> **Prompt:** Help me find a good name for this config setting:
> 
> // What should we do with pages in subdirectories when it comes to the main navigation menu?
> // You can choose between 'dropdown', 'flat', and 'hidden'.
> '***' => 'hidden', // Need a good name for this
> <attached full file with leading data removed>

 
**GPT 4o:** `subdirectory_navigation_handling`
**Claude:** `subdirectory_page_handling`

**Analysis:** We don't need to note that it includes navigation, and _behaviour_<sup>[sic]</sup> can be spelled different ways. Maybe we go for `subdirectory_handling`?

**Comment by caendesilva** [comment]
# Claude system analysis for the current Navigation API state

## Abstract 

The following is a report by Claude Sonnet with feedback on the new configuration structure and the overall Navigation API in its current state. /CDS

### Executive Summary


Based on the review, the navigation configuration and API for HydePHP appear to be well-designed and align closely with the project's philosophy. The system offers a good balance between simplicity and power, providing intuitive defaults while allowing for more complex customizations when needed.

The strengths of the system include its flexibility, intuitive design, automatic features, and good integration with Laravel conventions. The use of fluent interfaces and type-safe methods enhances the developer experience.

There are a few areas that could potentially be improved, such as documentation clarity, consistency in terminology, and API discoverability. However, these are relatively minor points in an otherwise solid system.

# HydePHP Navigation Configuration and API Feedback

## Overall Impression

The navigation configuration and API for HydePHP appear to be well-designed, flexible, and in line with the project's philosophy of simplicity, power, and quality. The system provides multiple ways to customize navigation, catering to different user preferences and use cases.

## Strengths

1. **Flexibility**: The system offers multiple ways to configure navigation (front matter, config files, API), allowing users to choose the most suitable method for their needs.

2. **Intuitive Design**: The configuration options are straightforward and use familiar concepts (labels, priorities, groups), making it easy for users to understand and implement.

3. **Automatic Features**: Features like automatic menu generation, subdirectory-based grouping, and numerical prefix ordering align well with the goal of reducing boilerplate and focusing on content.

4. **Progressive Disclosure**: The system provides simple defaults while allowing for more complex configurations when needed, adhering to the "Simplicity first. Power when you need it." motto.

5. **Laravel Integration**: The use of facades and configuration files will feel familiar to Laravel developers, enhancing the developer experience for that audience.

## Areas for Improvement

1. **Documentation Clarity**: While the provided documentation is comprehensive, it could benefit from more visual examples (e.g., code snippets alongside rendered menu examples) to help users better understand the outcomes of different configurations.

2. **Consistency in Terminology**: There are a few instances where multiple terms are used for the same concept (e.g., "group" and "category", "hidden" and "visible"). While flexibility is good, it might be beneficial to standardize on one term per concept to reduce potential confusion.

3. **API Discoverability**: The Navigation API seems powerful but might benefit from more prominence in the documentation, perhaps with a dedicated section or more examples of its use cases.

4. **Config File Structure**: The separation of navigation configuration between `hyde.php` and `docs.php` is logical but might be confusing for new users. Consider adding a clear explanation of this separation in the documentation.

## Specific Feedback

### Configuration Structure

1. The use of a fluent interface in the `hyde.php` config file is elegant and readable:

```php
'navigation' => Navigation::configure()
    ->setPagePriorities([...])
    ->setPageLabels([...])
    ->excludePages([...])
    ->addNavigationItems([...])
    ->setSubdirectoryDisplayMode('hidden'),
```

This approach aligns well with Laravel conventions and provides a clear, chainable way to configure navigation.

2. The `Navigation::item()` method for adding custom navigation items is intuitive and type-safe, which is great for developer experience.

3. The use of route keys for identifying pages is a smart choice, as it provides a consistent way to reference pages across different configuration methods.

### Navigation API

1. The separation of `MainNavigationMenu` and `DocumentationSidebar` classes, both extending a base `NavigationMenu` class, is a clean architectural choice. It allows for shared functionality while accommodating specific needs of each menu type.

2. The `NavigationItem` and `NavigationGroup` classes provide a flexible structure for representing menu items and dropdowns.

3. The ability to interact with navigation programmatically through the Navigation API is a powerful feature that aligns well with the "Power when you need it" part of the motto.

## Suggestions

1. Consider adding a visual configuration builder tool to the HydePHP ecosystem. This could help users, especially those less familiar with PHP, to easily create and visualize their navigation structure.

2. Explore the possibility of adding more advanced features to the Navigation API, such as conditional menu items or dynamic menu generation based on content tags or categories.

3. Consider providing more pre-built navigation templates or "recipes" for common site structures to further reduce boilerplate for users.

4. Enhance the documentation with more real-world examples and use cases, particularly for the Navigation API, to help users understand when and how to leverage its more advanced features.

5. Consider adding validation or type-checking for configuration values to catch potential errors early and provide helpful error messages.

In conclusion, the navigation configuration and API for HydePHP demonstrate a well-thought-out system that balances simplicity with power. It aligns well with the project's philosophy and provides a solid foundation for users to create and customize their site navigation.

**Comment by caendesilva** [comment]
Think we may actually want the `setSubdirectoryDisplayMode` to be dropdown by default.

**Comment by caendesilva** [comment]
Still work that needs to be done, but to be honest, I need a break from navigation.
